geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/RangeHistogram.scala

```diff
@@ -0,0 +1,219 @@
+          bucketIndex = numBins - 1
+
+        getBinKey(binSize, bucketIndex, lowerEndpoint)
+      } else {
+        null
+      }
+    }
+
+    override def getBinKey(binSize: java.lang.Integer, bucketIndex: Int, lowerEndpoint: lang.Integer): lang.Integer = {
+      lowerEndpoint + (binSize * bucketIndex)
+    }
+  }
+
+  implicit object BinAbleDouble extends BinAble[java.lang.Double] {
+    override def getBinSize(numBins: Int, lowerEndpoint: java.lang.Double, upperEndpoint: java.lang.Double): java.lang.Double = {
+      (upperEndpoint - lowerEndpoint) / numBins
+    }
+
+    override def getBinIndex(attributeValue: lang.Double,
+                             binSize: java.lang.Double,
+                             numBins: Int,
+                             lowerEndpoint: lang.Double,
+                             upperEndpoint: lang.Double): lang.Double = {
+      if (attributeValue >= lowerEndpoint && attributeValue < upperEndpoint) {
+        var bucketIndex = (attributeValue - lowerEndpoint) / binSize
+        if (bucketIndex >= numBins)
+          bucketIndex = numBins - 1
+
+        getBinKey(binSize, bucketIndex.toInt, lowerEndpoint)
+      } else {
+        null
+      }
+    }
+
+    override def getBinKey(binSize: java.lang.Double, bucketIndex: Int, lowerEndpoint: lang.Double): lang.Double = {
+      lowerEndpoint + (binSize * bucketIndex)
+    }
+  }
+
+  implicit object BinAbleFloat extends BinAble[java.lang.Float] {
+    override def getBinSize(numBins: Int, lowerEndpoint: java.lang.Float, upperEndpoint: java.lang.Float): java.lang.Float = {
+      (upperEndpoint - lowerEndpoint) / numBins
+    }
+
+    override def getBinIndex(attributeValue: lang.Float,
+                             binSize: java.lang.Float,
+                             numBins: Int,
+                             lowerEndpoint: lang.Float,
+                             upperEndpoint: lang.Float): lang.Float = {
+      if (attributeValue >= lowerEndpoint && attributeValue < upperEndpoint) {
+        var bucketIndex = (attributeValue - lowerEndpoint) / binSize
+        if (bucketIndex >= numBins)
+          bucketIndex = numBins - 1
+
+        getBinKey(binSize, bucketIndex.toInt, lowerEndpoint)
+      } else {
+        null
+      }
+    }
+
+    override def getBinKey(binSize: java.lang.Float, bucketIndex: Int, lowerEndpoint: lang.Float): lang.Float = {
+      lowerEndpoint + (binSize * bucketIndex)
+    }
+  }
+}
+
+import org.locationtech.geomesa.utils.stats.BinHelper._
+
+/**
+ * The range histogram's state is stored in a hashmap, where the keys are the bins and the values are the counts.
+ * A bin is a range of values but is treated as the lower endpoint of the bin to persist in the hashmap.
+ * A value is in a particular bin if it is equal to the bin value or less than the subsequent bin value.
+ *
+ * e.g. a range of 0 to 3 with 3 bins will result in these bins: [0, 1), [1, 2), [2, 3) and the hashmap will contain
+ * keys 0, 1, and 2.
```

Comment:
Yeah, the bins can't always be the same size if they're discrete values. don't think there's any way around that. But is that a problem? Seems like it would be better to have some size 1, some size 2, than to have all size 1 except one of size 3.